File: battle/attack/AtkModelEnemy.java
package common.battle.attack;

import common.CommonStatic;
import common.battle.data.DataEnemy;
import common.battle.entity.EEnemy;
import common.battle.entity.EUnit;
import common.battle.entity.EntCont;
import common.battle.entity.Entity;
import common.pack.Identifier;
import common.util.Data.Proc.SUMMON;
import common.util.unit.AbEnemy;
import common.util.unit.EForm;
import common.util.unit.Unit;
import org.jcodec.common.tools.MathUtil;

public class AtkModelEnemy extends AtkModelEntity {

	private final Proc[] cursed;

	protected AtkModelEnemy(EEnemy ent, float d0) {
		super(ent, d0, 1);
		String[] arr = { "KB", "STOP", "SLOW", "WEAK", "WARP", "CURSE", "SNIPER", "SEAL", "POISON", "BOSS",
				"POIATK", "ARMOR", "SPEED", "LETHARGY", "DMGCUT", "DMGCAP", "DELAY" };
		cursed = new Proc[data.getAtkCount()];
		for (int i = 0; i < cursed.length; i++) {
			cursed[i] = data.getAtkModel(i).getProc().clone();
			for (String s0 : arr)
				cursed[i].get(s0).clear();
		}
	}

	@Override
	public void summon(SUMMON proc, Entity ent, Object acs, int resist) {
		if (resist < 100) {
			if (proc.id == null || AbEnemy.class.isAssignableFrom(proc.id.cls)) {
				AbEnemy ene = Identifier.getOr(proc.id, AbEnemy.class);
				SUMMON.TYPE conf = proc.type;

				if (conf.same_health && ent.health <= 0)
					return;

				int time = proc.time;
				int allow = b.st.data.allow(b, ene);

				if (allow >= 0 || conf.ignore_limit) {
					int dis = b.getValueBetween(proc.dis, proc.max_dis);
					float ep = ent.pos + getDire() * dis;
					float mula = proc.mult * 0.01f;
					float mult = proc.mult * 0.01f;

					if (!conf.fix_buff) {
						mult *= (float) ((EEnemy) e).mult;
						mula *= (float) ((EEnemy) e).mula;
					}

					mula *= (100.0f - resist) / 100;
					mult *= (100.0f - resist) / 100;

					int layer = e.spawnLayer;

					if (proc.type.layer_type == CommonStatic.LayerType.SET)
						layer = b.getValueBetween(proc.min_layer, proc.max_layer);
					else if (proc.type.layer_type == CommonStatic.LayerType.RELATIVE)
						layer += b.getValueBetween(proc.min_layer, proc.max_layer);

					EEnemy ee = ene.getEntity(b, acs, mult, mula, layer, layer, 0, -1);

					ee.group = allow;
					if (proc.tba == -1)
						ee.setWaitTime(ee.data.getTBA());
					else if (proc.tba > 0)
						ee.setWaitTime(proc.tba);

					if (ep < ee.data.getWidth())
						ep = ee.data.getWidth();
					if (ep > b.st.len - 800)
						ep = b.st.len - 800;

					ee.added(1, (int) ep);

					b.tempe.add(new EntCont(ee, time));

					if (conf.same_health)
						ee.health = e.health;

					ee.setSummon(conf.anim_type, conf.bond_hp ? e : null);
				}
			} else {
				Unit u = Identifier.getOr(proc.id, Unit.class);
				SUMMON.TYPE conf = proc.type;

				if (conf.same_health && ent.health <= 0)
					return;

				int time = proc.time;

				if (u.forms.length == 0)
					return;

				int form = Math.max(0, Math.min(proc.form - 1, u.forms.length - 1));

				if (b.entityCount(-1) < b.maxNum - u.forms[form].du.getWill() || conf.ignore_limit) {
					int lvl = proc.mult;
					lvl = MathUtil.clip(lvl, 1, u.max + u.maxp);
					lvl = (int) (lvl * (100.0 - resist) / 100);

					int dis = b.getValueBetween(proc.dis, proc.max_dis);
					float up = ent.pos + getDire() * dis;
					int layer = e.spawnLayer;

					if (proc.type.layer_type == CommonStatic.LayerType.SET)
						layer = b.getValueBetween(proc.min_layer, proc.max_layer);
					else if (proc.type.layer_type == CommonStatic.LayerType.RELATIVE)
						layer += b.getValueBetween(proc.min_layer, proc.max_layer);

					EForm ef = new EForm(u.forms[form], lvl);
					EUnit eu = ef.invokeEntity(b, lvl, layer, layer);
					if (conf.same_health)
						eu.health = e.health;

					eu.added(-1, (int) up);
					b.tempe.add(new EntCont(eu, time));
					eu.setSummon(conf.anim_type, conf.bond_hp ? e : null);
				}
			}
		} else
			ent.anim.getEff(INV);
	}

	@Override
	protected int getAttack(int ind, Proc proc) {
		int atk = atks[ind];
		extraAtk(ind);
		if (abis[ind] == 1)
			setProc(ind, proc);
		if (e.data instanceof DataEnemy)
			for (int j : BCShareable) proc.getArr(j).set(e.getProc().getArr(j));

		return atk;
	}

	@Override
	public Proc getProc(int ind) {
		if (e.status[P_CURSE][0] > 0 && e.status[P_SEAL][0] == 0 && ind < cursed.length)
			return cursed[ind];
		return super.getProc(ind);
	}

}

File: battle/attack/AtkModelEnemy.kt
package common.battle.attack

import common.battle.entity.EEnemy
import common.battle.entity.EntCont
import common.battle.entity.Entity
import common.util.Data
import common.util.Data.Proc.SUMMON
import common.util.unit.AbEnemy

class AtkModelEnemy(ent: EEnemy, d0: Double) : AtkModelEntity(ent, d0) {
    private val cursed: Array<Proc?>
    override fun summon(proc: SUMMON, ent: Entity, acs: Any?) {
        val ene: AbEnemy = proc.id.get() as AbEnemy
        val conf: Data.Proc.SUMMON.TYPE = proc.type
        val time: Int = proc.time
        val allow: Int = b.st.data.allow(b, ene)
        if (ene != null && (allow >= 0 || conf.ignore_limit)) {
            var ep: Double = ent.pos + getDire() * proc.dis
            var mula: Double = proc.mult * 0.01
            var mult: Double = proc.mult * 0.01
            if (!conf.fix_buff) {
                mult *= (e as EEnemy).mult
                mula *= (e as EEnemy).mula
            }
            var l0 = 0
            var l1 = 9
            if (!conf.random_layer) {
                l1 = e.layer
                l0 = l1
            }
            val ee: EEnemy = ene.getEntity(b, acs, mult, mula, 0, l0, l1)
            ee.group = allow
            if (ep < ee.data.getWidth()) ep = ee.data.getWidth().toDouble()
            if (ep > b.st.len - 800) ep = b.st.len - 800.toDouble()
            ee.added(1, ep.toInt())
            b.tempe.add(EntCont(ee, time))
            if (conf.same_health) ee.health = e.health
            ee.setSummon(conf.anim_type)
        }
    }

    protected override fun getAttack(ind: Int, proc: Proc): Int {
        var atk: Int = atks.get(ind)
        extraAtk(ind)
        if (abis.get(ind) == 1) setProc(ind, proc)
        if (e.status.get(Data.Companion.P_WEAK).get(0) > 0) atk = atk * e.status.get(Data.Companion.P_WEAK).get(1) / 100
        if (e.status.get(Data.Companion.P_STRONG).get(0) != 0) atk += atk * e.status.get(Data.Companion.P_STRONG).get(0) / 100
        return atk
    }

    protected override fun getBaseAtk(ind: Int): Int {
        return atks.get(ind)
    }

    protected override fun getProc(ind: Int): Proc? {
        return if (e.status.get(Data.Companion.P_CURSE).get(0) > 0 && e.status.get(Data.Companion.P_SEAL).get(0) == 0) cursed[ind] else super.getProc(ind)
    }

    init {
        val arr = arrayOf("KB", "STOP", "SLOW", "WEAK", "WARP", "CURSE", "SNIPER", "SEAL", "POISON", "BOSS", "IMUATK",
                "POIATK")
        cursed = arrayOfNulls<Proc>(data.getAtkCount())
        for (i in cursed.indices) {
            cursed[i] = data.getAtkModel(i).getProc().clone()
            for (s0 in arr) cursed[i].get(s0).clear()
        }
    }
}
